drivers/gpio/gpio_tca9538.c

```diff
@@ -0,0 +1,524 @@
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	/* Store previous input state then read new value */
+	previous_state = drv_data->pin_state.input;
+	rc = update_input(dev);
+	if (rc != 0) {
+		goto out;
+	}
+
+	/* Find out which pins have changed state */
+	current_state = drv_data->pin_state.input;
+	transitioned_pins = previous_state ^ current_state;
+
+	/* Mask gpio transactions with rising/falling edge interrupt config */
+	interrupt_mask = (irq_state->rising & transitioned_pins &
+			  current_state);
+	interrupt_mask |= (irq_state->falling & transitioned_pins &
+			   previous_state);
+
+out:
+	k_sem_give(&drv_data->lock);
+
+	if (rc == 0) {
+		gpio_fire_callbacks(&drv_data->cb, dev, interrupt_mask);
+	}
+
+	return rc;
+}
+
+/**
+ * @brief Work handler for TCA9538 interrupt
+ *
+ * @param work Work struct that contains pointer to interrupt handler function
+ */
+static void gpio_tca9538_work_handler(struct k_work *work)
+{
+	struct tca9538_drv_data *drv_data =
+		CONTAINER_OF(work, struct tca9538_drv_data, work);
+
+	gpio_tca9538_handle_interrupt(drv_data->dev);
+}
+
+/**
+ * @brief ISR for intterupt pin of TCA9538
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param gpio_cb Pointer to callback function struct
+ * @param pins Bitmask of pins that triggered interrupt
+ */
+static void gpio_tca9538_init_cb(const struct device *dev,
+				 struct gpio_callback *gpio_cb, uint32_t pins)
+{
+	struct tca9538_drv_data *drv_data =
+		CONTAINER_OF(gpio_cb, struct tca9538_drv_data, gpio_cb);
+
+	ARG_UNUSED(pins);
+
+	k_work_submit(&drv_data->work);
+}
+
+static int gpio_tca9538_config(const struct device *dev, gpio_pin_t pin,
+			       gpio_flags_t flags)
+{
+	const struct tca9538_config *cfg = dev->config;
+	struct tca9538_drv_data *drv_data = dev->data;
+	struct tca9538_pin_state *pins = &drv_data->pin_state;
+	int rc = 0;
+	bool data_first = false;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	/* Zephyr currently defines drive strength support based on
+	 * the behavior and capabilities of the Nordic GPIO
+	 * peripheral: strength defaults to low but can be set high,
+	 * and is controlled independently for output levels.
+	 *
+	 * The TCA9538 supports only high strength, and does not
+	 * support different strengths for different levels.
+	 *
+	 * Until something more general is available reject any
+	 * attempt to set a non-default drive strength.
+	 */
+	if ((flags & (GPIO_DS_ALT_LOW | GPIO_DS_ALT_HIGH)) != 0) {
+		return -ENOTSUP;
+	}
+
+	/* Open drain and open source not supported */
+	if ((flags & GPIO_OPEN_DRAIN) != 0 || (flags & GPIO_OPEN_SOURCE) != 0) {
+		return -ENOTSUP;
+	}
+
+	/* The TCA9538 has no internal pull up support */
+	if (((flags & GPIO_PULL_UP) != 0) || ((flags & GPIO_PULL_DOWN) != 0)) {
+		return -ENOTSUP;
+	}
+
+	/* The TCA9538 does not support internal debounce */
+	if ((flags & GPIO_INT_DEBOUNCE) != 0) {
+		return -ENOTSUP;
+	}
```

Comment:
According to the documentation
> Drivers that do not support a debounce feature should ignore this flag rather than rejecting the configuration with -ENOTSUP

Reply:
Removed this check.